doc/adr007-flattening.md

```diff
@@ -0,0 +1,255 @@
+```bluespec
+module A {
+  const N: int
+  val a = N
+}
+
+module B {
+  import A(N=1) as A1
+  export A1
+}
+```
+
+Currently, this prevents us from using `A1` without a namespace, since we can't write an export for `import A(N=1).*`. @konnov suggested the introduction of `export *` for this end, which is still in consideration.
+
+#### `importedFrom`
+
+This is a simple reference to the import/export/instance statement that created a definition, also gathered during name collection. Used in flattening to distinguish definitions that come from instances from definitions that come from imports/exports.
+
+### Instance Flattener
+
+The instance flattener gets rid of instances, which is the most complicated part of flattening. We cannot use the same id for the same definition of different instances, as they might assume different values[^1].
+
+Each instance statement is replaced with an import statement and generates a new module. The new module has the constants replaced with `pure val` definitions (from the overrides in the instance statement), and all its names are fully-qualified, constructed with the previously defined namespace algorithm. The new import statement, replacing the instance statement, is a unqualified import of the new module.
+
+The names in the module containing the instance statement also have to be updated with a namespace, since they might refer to definitions that now include a namespace. We look at each name in the module and, if it refers to a definition that comes from an instance, (that is, its `importFrom` field has an instance), then we replace the name with a namespaced version of it (using the `namespaces` field). For this to work properly, any existing `import` and `export` statements should have been flattened before the instance flattener is run, ensuring that the definitions referred to by these names already have their proper namespaces.
+
+All of this tinkering on names and the addition of a matching import statement makes it so the module outputted by the instance flattener can be name resolved. The name resolution of this module will produce a lookup table where the names referring to instances point to the new modules (produced for each instance), which are no longer instances. Therefore, after instance flattening, we have instance-less modules and their corresponding lookup table.
+
+#### Example
+
+Take this module as example, where `A` is instanced twice with the same name:
+```bluespec
+module A {
+  const N: int // id 1
+  val a = N // id 2
+}
+
+module B {
+  import A(N=1) as A1
+  val b = A1::a // id 3
+}
+
+module C {
+  import A(N=0) as A1
+  val C = A1::a // id 4
+}
+```
+
+The instance flattener will create two new modules: `B::A1` and `C::A1`, with new ids for the definitions inside it.
+```bluespec
+module A {
+  const N: int // id 1
+  val a = N // id 2
+}
+
+module B::A1 {
+  pure val B::A1::N = 1 // id 5
+  pure val B::A1::a = B::A1::N // id 6
```

Comment:
Maybe I don't have full context on how we resolve qualified names –
is the FQN of `N` `B::A1::N` or `B::A1::B::A1::N`?

Reply:
In theory it is the latter, but we don't actually use FQN for anything right now or have anything like that in the implementation.

Reply:
First of all, the module names need to be qualified in order to differentiate modules generated for different instances (i.e. `import A(N=1).*` in module `Foo` vs `import A(N=2).*` in module `Bar`, etc). So that needs to be qualified for sure.

As for the definitions: In theory (as in I haven't tried this fully to know that aren't more complications than I see), we could avoid qualifying the definition names here. We do qualify them to simplify the process of guaranteeing no conflicts in the flattened module. That's because these definitions will be imported later (as the instance statement was replaced with an import one), and this approach with the qualified names makes this very straight forward because the `namespaces` field doesn't need further manipulation.

If we didn't qualify the definitions here, we would also have to change the import statement produced from the instance statement, and that would make it so the `namespaces` field generated from the instance statement (during name collection) is no longer compatible with the new import statement.

My approach: `import A(N=1).*` becomes `import myModule::A.*` (no namespaces need to be added in neither)
Other approach: `import A(N=1).*` becomes `import myModule::A` (qualified import). We would have to update the `namespaces` field for the definitions inside `myModule::A` to contain `['A', 'myModule']`. This way, when the import statement is processed by the last flattening round, the flattener will add the namespaces and the definitions in the flattened module would be qualified like `myModule::A::foo`.

The end result in the flattened module is the same. Writing this down makes me realize that this other approach makes more sense than I initially thought. I think I'll try it out and see if it makes sense to go with it (in which case I'll update the ADR). Thanks for pointing it out.

PS: There is additional complexity to these approaches when the instance is qualified i.e. (`import A(N=1) as A1`), so I picked a simpler scenario to explain. I actually think the "Other approach" might be better for reducing that complexity.

Reply:
Thanks for walking me through this! 🙏

Looks like neither way is completely friction-less, so picking the one easiest to implement makes a lot of sense!
Let's update the ADR with a note once we know which way we went – seems like a decision that's good to persist for future reference :)